### PR TITLE
Add closed-form MXFP4 -> NVFP4 weight cast (--cast_mxfp4_to_nvfp4)

### PR DESCRIPTION
**Pull request by cjluo-nv**
## Summary

- New `--cast_mxfp4_to_nvfp4` flag in `hf_ptq.py` (and `huggingface_example.sh`) that converts an MXFP4 source checkpoint (e.g. `openai/gpt-oss-20b`) into an NVFP4 export with **bit-exact** weight reconstruction for the in-range blocks.
- The cast pins NVFP4's `scale_2 = 2^m` (where `m = k_max − 8`) and `_amax = 6·2^k_j` per NVFP4 block, both read from the source `*_scales`. The resulting per-block scale `2^(k_j − m)` is exactly representable in E4M3, so `round_to_E2M1(value / 2^k_j)` yields the original MXFP4 nibble verbatim. For out-of-range blocks (`k_max − k_j > 17`) the per-block amax falls back to data-derived `max(|w_block|)`, which keeps the post-E4M3-clamp scale close to the block's actual magnitude.

## Verification

End-to-end on `openai/gpt-oss-20b` with `--qformat=nvfp4_mlp_only --cast_mxfp4_to_nvfp4`:

```
[cast_mxfp4_to_nvfp4] overrode 48/48 weight quantizers
[cast_mxfp4_to_nvfp4] lossless layers: 48/48 (100.00%)
[cast_mxfp4_to_nvfp4] lossless blocks: 597196800/597196800 (100.0000%)
```

End-to-end on `openai/gpt-oss-120b` with the same flags (4×B200, `--use_seq_device_map --gpu_max_mem_percentage 0.5 --calib_batch_size 4`):

```
[cast_mxfp4_to_nvfp4] overrode 72/72 weight quantizers
[cast_mxfp4_to_nvfp4] lossless layers: 67/72 (93.06%)
[cast_mxfp4_to_nvfp4] lossless blocks: 3583179586/3583180800 (100.0000%)
```

Five layers fall into the OOR regime (block-spread > 17); the remaining 1,214 OOR blocks use the data-derived per-block amax fallback. Block-level losslessness is **99.99996%** end-to-end.

Per-tensor MSE between MXFP4 source dequant and NVFP4 export dequant (~19B elements):

| Metric | Without cast | With cast |
|---|---|---|
| Per-tensor SNR | ~26.4 dB (FP4 noise floor) | **∞ (every tensor)** |
| Total RMSE | 8.67e−02 | **0** |
| max\|err\| | up to 8.0e+1 | **0** |

## Modelopt-side enablers

- `max_calibrate` auto-promotes static-block NVFP4 weight quantizers to `NVFP4StaticQuantizer` at the end of calibration.
- `static_blockwise_fp4_fake_quant` kernel accepts N-D inputs (was 2D-only), unblocking MoE expert weights of shape `(E, F, K)`.
- BMM-experts NVFP4 export routes through `get_weights_scaling_factor_from_quantizer` for static-mode quantizers, so the pinned `_amax` is actually consumed.
- `set_expert_quantizer_amax` scalar-reduces per-quantizer amax before stacking, supporting per-block (vs scalar) static-mode amax.

## Test plan

- [x] Unit tests at `tests/examples/llm_ptq/test_cast_mxfp4_to_nvfp4.py` (15 tests, all passing) cover: scalar/global-amax math, per-block hybrid (in-range closed-form vs OOR data-derived), shape preservation, key collection, and end-to-end `build_amax_map` against a synthetic safetensors checkpoint.
- [x] End-to-end PTQ → export on `openai/gpt-oss-20b` (`nvfp4_mlp_only` qformat) with `--cast_mxfp4_to_nvfp4` succeeds; export takes ~21 s. 100% lossless cast (48/48 layers, 597,196,800 / 597,196,800 blocks).
- [x] End-to-end PTQ → export on `openai/gpt-oss-120b` (4×B200, `nvfp4_mlp_only`, `--use_seq_device_map --gpu_max_mem_percentage 0.5 --calib_batch_size 4`). 67/72 layers fully lossless; 99.99996% block-level losslessness (3,583,179,586 / 3,583,180,800).
- [x] TRT-LLM serving validation (TRT-LLM 1.3.0rc11, B200) on both exported NVFP4 checkpoints via `examples/llm_ptq/run_tensorrt_llm.py`:
  - **20b** (TP=1): 18.3 GB GPU memory; coherent generation. Sample: *"Quantum computing is poised to revolutionize data analysis. However, its potential is currently limited by quantum hardware constraints, including error rates, qubit lifetimes, and lack of fault tolerance…"*
  - **120b** (TP=4): 36.4 GB / GPU; coherent generation. Sample: *"Quantum computing is poised to revolutionize data storage and processing. These rare earth-based systems could serve as robust qubits; resistant to environmental decoherence…"*
- [x] MSE comparison script (run separately during development) confirms per-tensor SNR=∞ across all 48 MoE expert tensors.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added an MXFP4→NVFP4 weight-format cast utility, a CLI flag to enable it, and shell/parser helpers to invoke it; integrates into the PTQ workflow and enforces NVFP4-family qformat constraints.

* **Bug Fixes**
  * Fixed static NVFP4 export for expert weights.
  * Improved collection/handling of quantizer amax values to avoid shape issues.
  * Generalized FP4 kernel to accept flexible tensor dimensionality.
  * Ensured static-block NVFP4 promotion during calibration.

* **Tests**
  * Added comprehensive tests covering the conversion workflow, helpers, and end-to-end application.

* **Documentation**
  * Updated README and changelog with usage, constraints, and GPT-OSS support.

### CODE REVIEW

File: examples/llm_ptq/cast_mxfp4_to_nvfp4.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+def _collect_scales_keys(ckpt_dir: Path) -> dict[str, Path]:
+    """Return ``{tensor_name: shard_path}`` for every ``*_scales`` key."""
+    return _collect_keys_with_suffix(ckpt_dir, "_scales")
+
+
+def build_amax_map(checkpoint_dir: str | Path) -> dict[str, dict]:
+    """Walk the source MXFP4 checkpoint and build the per-layer amax map.
+
+    Args:
+        checkpoint_dir: Path to a Hugging Face checkpoint directory whose
```
Comment (meenchen) [inline examples/llm_ptq/cast_mxfp4_to_nvfp4.py:253]: > _Bot comment._

**Resource leak**: `safe_open` handles in `handles` dict are never closed. Consider using a `try/finally` block or closing them at the end:
```python
try:
    # ... loop body ...
finally:
    for h in handles.values():
        # safe_open objects are context managers
        pass  # they get GC'd, but explicit cleanup is better
```
Same issue exists in `apply_to_model` at the other `handles` dict (line ~273).

File: examples/llm_ptq/cast_mxfp4_to_nvfp4.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    # Python int. All-zero scales fall through with the -E8M0_BIAS sentinel,
+    # which leaves every block trivially in-range (closed_form_ideal == 0 there).
+    nonzero = e8m0_scales > 0
+    sentinel = torch.full_like(k, -E8M0_BIAS)
+    k_max = torch.where(nonzero, k, sentinel).amax()
+    delta = k - (k_max - E4M3_KMAX)
+    in_range = (delta >= E4M3_KMIN) & (delta <= E4M3_KMAX)
+
+    # Fast path: if every block fits E4M3's [-9, 8] window the per-block amax
+    # is just the closed-form ideal, and we can skip the per-byte nibble scan
```
Comment (meenchen) [inline examples/llm_ptq/cast_mxfp4_to_nvfp4.py:187]: > _Bot comment._

This `assert` validates user-provided checkpoint data (the key suffix). If someone passes a malformed key, this would be silently skipped under `python -O`. Since this is input validation rather than a programming invariant, consider using a `ValueError` instead:
```python
if not blocks_key.endswith("_blocks"):
    raise ValueError(f"Unexpected key {blocks_key!r}")
```

File: tests/examples/llm_ptq/test_cast_mxfp4_to_nvfp4.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+
+# ---------- magnitude table cache ------------------------------------------
+
+
+def test_e2m1_magnitude_table_cached_per_device():
+    t1 = cast._e2m1_magnitude_table(torch.device("cpu"))
+    t2 = cast._e2m1_magnitude_table(torch.device("cpu"))
+    assert t1 is t2  # cached: same object
+    assert t1.tolist() == cast._E2M1_MAGNITUDE
```
Comment (meenchen) [inline tests/examples/llm_ptq/test_cast_mxfp4_to_nvfp4.py:282]: > _Bot comment._

**Missing test coverage for `apply_to_model`**: The most complex and user-facing function in the module is not tested. It has significant logic around matching quantizer names to model modules, asserting quantizer types, reshaping amax buffers, and writing to quantizer state. Consider adding at least one test with a mock model (e.g., a simple `nn.Module` with a mock `NVFP4StaticQuantizer` child) to exercise the happy path and verify that `global_amax` and `_amax` are written correctly.

File: examples/llm_ptq/cast_mxfp4_to_nvfp4.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+        blocks_key = tensor_key[: -len("_scales")] + "_blocks"
+        qname = quantizer_name_from_blocks_key(blocks_key)
+        amax_map[qname] = {"global_amax": global_amax, **info}
+
+    return amax_map
+
+
+def apply_to_model(
+    model: "torch.nn.Module",
```
Comment (meenchen) [inline examples/llm_ptq/cast_mxfp4_to_nvfp4.py:312]: > _Bot comment._

Same concern as `quantizer_name_from_blocks_key`: this `assert` validates runtime data (whether a paired `_blocks` tensor exists in the checkpoint). A missing pair is a data error, not a programming bug. Consider `ValueError` or a descriptive `RuntimeError`.

**Review by meenchen — COMMENTED** [review]
> _Bot review — DM the bot to share feedback._

Well-documented PR with solid math and good test coverage for the core conversion logic. The supporting fixes across the codebase (kernel flexibility, export path, expert amax stacking) are small and correct. A few issues to address:

1. **Resource leak**: `safe_open` handles are stored in dicts but never closed in both `build_amax_map` and `apply_to_model`.
2. **Missing test for `apply_to_model`**: The most complex function in the module has no unit tests. The test suite covers the math functions and checkpoint reading well, but the model-interaction path (the function users actually call) is untested.
3. **Minor**: `quantizer_name_from_blocks_key` uses `assert` for input validation on user-provided checkpoint data — this should be a proper exception since malformed checkpoint keys are a runtime data condition, not a programming invariant.

**Review by meenchen — COMMENTED** [review]
> _Bot review — DM the bot to share feedback._

Re-review: All critical comments from the previous review have been addressed.

**Addressed:**
- Resource leak: `safe_open` handles now managed via `_shard_reader()` context manager with `ExitStack` — deterministic cleanup on all paths.
- NVFP4StaticQuantizer promotion: moved before the `distributed_sync` early return in `max_calibrate`, so single-process callers also get promotion.
- Block trailing dim validation: now checks `blocks.shape[-1] != 16`.
- I/O skip for lossless layers: fully-lossless layers compute per-block amax from scales alone, avoiding the 16x larger block read.
- Meta tensor guard: `existing_amax.is_meta` check added in `set_expert_quantizer_amax`.
- `apply_to_model` test coverage: 3 tests added (happy path, missing blocks pair, wrong quantizer type).
- Fail-fast validation: `--cast_mxfp4_to_nvfp4` now rejects non-NVFP4 qformats and `auto_quantize_bits`.

**Design review**: This is a focused conversion utility, not a new architectural subsystem. The 6-directory span comes from small targeted changes across existing subsystems (kernel N-D input support, export static-quantizer routing, calibration promotion timing) plus the new conversion script in `examples/`. The design is appropriate and well-scoped.

**Code quality**: Well-documented with thorough docstrings explaining the math. Test suite covers core math functions, edge cases (zero blocks, OOR blocks), checkpoint I/O, and the model-interaction path. The ~950 lines are cohesive and can't be reasonably split further. License headers match the canonical `LICENSE_HEADER`.

Complex PR: spans 6 directories (≥ 5). Looping in a human for approval.

File: examples/llm_ptq/scripts/huggingface_example.sh [hunk 5]
```diff
@@ -143,6 +143,10 @@ if [ -n "$MOE_CALIB_EXPERTS_RATIO" ]; then
     PTQ_ARGS+=" --moe_calib_experts_ratio=$MOE_CALIB_EXPERTS_RATIO "
 fi
 
+if $CAST_MXFP4_TO_NVFP4; then
+    PTQ_ARGS+=" --cast_mxfp4_to_nvfp4 "
```
Comment (meenchen) [inline examples/llm_ptq/scripts/huggingface_example.sh:147]: Should we make this a default option? I am thinking of detecting MXFP4 quant config on the fly, and just applying the casting by default.

**Comment by realAsma** [comment]
> For out-of-range blocks (k_max - k_j > 17) the per-block amax falls back to data-derived max(|w_block|)

Is this correct? Out of range blocks will be `k_j - k_max > 16` or `k_j - k_max < -15` (which is `k_max - k_j > 15`).
Please let me know what I am missing.

Update: I understand now that this computation is using E4M3 subnormal values. `torch.float8_e4m3fn` can exactly represent powers of two down to `2^-9`, not only the normal range. So the lossless range for the scale exponent is `-9 <= k_j - m <= 8`; with `m = k_max - 8`, this becomes `k_max - k_j <= 17`. That makes the PR condition `k_max - k_j > 17` for out-of-range blocks correct.

File: examples/llm_ptq/scripts/huggingface_example.sh (hunk 5 above)
Comment (cjluo-nv) [reply examples/llm_ptq/scripts/huggingface_example.sh:147]: So far we only have gpt-oss and DSV4 (I will update another code path in the deepseek example). I think it's ok to have this flag for now. Also each model may have mxfp4 implemented differently.

**Review by meenchen — APPROVED** [review]

File: examples/llm_ptq/hf_ptq.py [hunk 6]
```diff
@@ -1087,6 +1088,21 @@ def quantize_main(
...
+        # picks them up automatically at the end of max_calibrate.
+        if args.cast_mxfp4_to_nvfp4:
+            quant_cfg = copy.deepcopy(quant_cfg)
+            for entry in quant_cfg.get("quant_cfg", []):
+                qname = entry.get("quantizer_name", "")
+                cfg = entry.get("cfg") or {}
+                bs = cfg.get("block_sizes")
+                if "weight_quantizer" in qname and isinstance(bs, dict):
+                    bs = {**bs, "type": "static"}
+                    entry["cfg"] = {**cfg, "block_sizes": bs}
```
Comment (realAsma) [inline examples/llm_ptq/hf_ptq.py]: can we move these to a helper method?

File: modelopt/torch/quantization/model_calib.py [hunk 7]
```diff
@@ -163,6 +163,15 @@ def max_calibrate(
...
             module.layer_sync_moe_local_experts_amax(sync_weight_amax=sync_expert_weight_amax)
 
+    # Promote eligible static-block NVFP4 weight quantizers to NVFP4StaticQuantizer
+    # so the static blockwise fake-quant path is used in forward and the export
+    # picks up the two-level (per-block + global) scaling. Run before the
+    # ``distributed_sync`` early return so single-process callers also get the
+    # promotion. ``promote_nvfp4_static_quantizers`` only promotes when
+    # ``is_static_block_quant`` is True and the per-block ``_amax`` buffer is
+    # populated, so it's a no-op for dynamic-block / non-NVFP4 configs.
+    promote_nvfp4_static_quantizers(model)
```
Comment (realAsma) [inline modelopt/torch/quantization/model_calib.py:173]: does this break GPTQ/MSE support?

Comment (cjluo-nv) [reply modelopt/torch/quantization/model_calib.py:173]: It should not. It just says: if the weight quantizer is static instead of dynamic fp4, let's make it static fp4 weight quantizer type instead of the parent tensor quantizer.

**Review by realAsma — APPROVED** [review]
